plugin/helpers_bots.go

```diff
@@ -18,11 +18,24 @@ func (p *HelpersImpl) EnsureBot(bot *model.Bot) (retBotId string, retErr error)
 	}
 
 	// If we fail for any reason, this could be a race between creation of bot and
-	// retreval from anouther EnsureBot. Just try the basic retrieve existing again.
+	// retreval from another EnsureBot. Just try the basic retrieve existing again.
 	defer func() {
 		if retBotId == "" || retErr != nil {
-			time.Sleep(time.Second)
-			botIdBytes, err := p.API.KVGet(BOT_USER_KEY)
+			var err error
+			var botIdBytes []byte
+
+			// this func is the operation that will be retried if
+			// returns not nil. It will be retried up to 3 times with
+			// and the wait it 128, 256 and 512 ms
+			retryOperation := func() error {
+				botIdBytes, err = p.API.KVGet(BOT_USER_KEY)
+				if err != nil {
+					return err
+				}
+				return nil
+			}
+			err = utils.ProgressiveRetry(retryOperation)
+
```

Comment:
Orthogonal to your changes, but would it make sense to update `retErr` with the current error?
```suggestion
if err != nil {
    retErr = err
    return
}
```

Reply:
Hmm, aren't we overwriting previous set errors then with a possible error from `KVGet`?

For example the error in the following code: `return "", errors.Wrap(createBotErr, "failed to create bot")` might be overwritten. By doing so we will never be able to handle the `creation` error from the calling function.

Would @lieut-data say this is an correct understanding?

Reply:
That's fair: if we assume the caller wanted the error that kickstarted the whole chain of failures, then yes, we should probably leave `retErr` alone.